qa/tasks/cephfs/test_acls.py

```diff
@@ -0,0 +1,480 @@
+        self.assert_dirs_are_accessible(testdirs, test_with_other_user,
+                                        expecting_access=False)
+
+    #######################################
+    #
+    # Test ACLs
+    #
+    ######################################
+
+    def test_acls_for_file_owner(self):
+        self.setup()
+        self.mount_a.run_as_root('chown %s:%s %s' % (self.currentuser, self.currentgroup, self.testdir1))
+        self.mount_a.run_shell('chmod u-rwx %s' % (self.testdir1))
+
+        # preliminary test
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+        self.mount_a.run_as_root(['setfacl', '-m', self.fo_acl, self.testdir1])
+        self.assert_acls_are_('present', self.fo_acl, self.testdir1)
+        self.assert_dirs_are_accessible(self.testdir1)
+
+    def test_acls_for_file_group(self):
+        self.setup()
+        self.mount_a.run_as_root('chown %s:%s %s' % (self.currentuser, self.currentgroup, self.testdir1))
+        self.mount_a.run_shell('chmod g-rwx %s' % (self.testdir1))
+
+        # preliminary test
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+        self.mount_a.run_as_root(['setfacl', '-m', self.fg_acl, self.testdir1])
+        self.assert_acls_are_('present', self.fg_acl, self.testdir1)
+        self.assert_dirs_are_accessible(self.testdir1)
+
+    def test_acls_for_other(self):
+        self.setup()
+        self.mount_a.run_as_root('chmod o-rwx %s' % (self.testdir1))
+
+        # preliminary test
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+        self.mount_a.run_as_root(['setfacl', '-m', self.o_acl, self.testdir1])
+        self.assert_acls_are_('present', self.o_acl, self.testdir1)
+        self.assert_dirs_are_accessible(self.testdir1)
+
+    def test_acls_for_current_user(self):
+        self.setup()
+
+        # preliminary test
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+        self.mount_a.run_as_root(['setfacl', '-m', self.u_acl, self.testdir1])
+        self.assert_acls_are_('present', self.u_acl, self.testdir1)
+        self.assert_dirs_are_accessible(self.testdir1)
+
+        self.mount_a.run_as_root(['setfacl', '-x', self.rm_u_acl, self.testdir1])
+        self.assert_acls_are_('absent', self.u_acl, self.testdir1)
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+    def test_acls_for_current_group(self):
+        self.setup()
+
+        # preliminary test
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+        self.mount_a.run_as_root('setfacl -m %s %s' % (self.g_acl, self.testdir1))
+        self.assert_acls_are_('present', self.g_acl, self.testdir1)
+        self.assert_dirs_are_accessible(self.testdir1)
+
+        self.mount_a.run_as_root(['setfacl', '-x', self.rm_g_acl, self.testdir1])
+        self.assert_acls_are_('absent', self.g_acl, self.testdir1)
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+    # XXX: Always fails
```

Comment:
Do we know why this fails? Where does it fall down?

Reply:
I get the "permission denied" error message -
```
2019-03-20 18:58:31,865.865 INFO:__main__:Running ['sudo', '-u', 'dummy', '-s', '/bin/bash', '-c', 'ls /tmp/tmpLsYBMs/mnt.0/dir1']
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
> /home/rishabh/repos/ceph/enable-acls-by-default/qa/tasks/cephfs/test_acls.py(357)_test_acl_for_other_user()
-> assert proc.returncode == 0
(Pdb) p proc.returncode
2
(Pdb) p proc.stderr().getvalues()
*** AttributeError: AttributeError('StringIO instance has no __call__ method',)
(Pdb) p proc.stderr().getvalue()
*** AttributeError: AttributeError('StringIO instance has no __call__ method',)
(Pdb) p proc.stderr.getvalue()
"ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied\n"
(Pdb) p output
'# file: tmp/tmpLsYBMs/mnt.0/dir1\n# owner: root\n# group: root\nuser::rwx\nuser:dummy:rwx\ngroup::---\nmask::rwx\nother::---\n\n'
(Pdb) 
```

 However, this is never the case with user using which I run the test -
```
rishabh$ ls /tmp/tmpLsYBMs/mnt.0/dir1
ls: cannot open directory '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
rishabh$ sudo ls /tmp/tmpLsYBMs/mnt.0/dir1
dir2
rishabh$ sudo -u dummy -s /bin/bash -c "ls /tmp/tmpLsYBMs/mnt.0/dir1"
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied

dummy$ ls /tmp/tmpLsYBMs/mnt.0/dir1
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
dummy$ sudo getfacl /tmp/tmpLsYBMs/mnt.0/dir1
getfacl: Removing leading '/' from absolute path names
# file: tmp/tmpLsYBMs/mnt.0/dir1
# owner: root
# group: root
user::rwx
user:dummy:rwx
group::---
mask::rwx
other::---
dummy$ logout

rishabh$ sudo setfacl -m user:rishabh:rwx /tmp/tmpLsYBMs/mnt.0/dir1
rishabh$ ls /tmp/tmpLsYBMs/mnt.0/dir1
dir2
```

I felt that the absence of keyring might be the reason, but that doesn't looks like the case - 
```
rishabh$ sudo cp ~/ceph/build/keyring /home/dummy/.ceph
rishabh$ sudo diff ~/ceph/build/keyring /home/dummy/.ceph/keyring
rishabh$ sudo -u dummy -s /bin/bash -c "ls /tmp/tmpLsYBMs/mnt.0/dir1"
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
rishabh$ sudo su - dummy
dummy$ ls /tmp/tmpLsYBMs/mnt.0/dir1
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
dummy$ 

```

Reply:
This happens not only for a user in some other group but also for users in the same group -

```
rishabh$ sudo gpasswd -a user2 user2
Adding user user2 to group user2
rishabh$ groups user2
user2 : user2 rishabh
rishabh$ sudo -u user2 -s /bin/bash -c "ls /tmp/tmpLsYBMs/mnt.0/dir1"
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/dir1': Permission denied
rishabh$ sudo su - user2
user2$ ll /tmp/tmpLsYBMs/mnt.0/
ls: cannot access '/tmp/tmpLsYBMs/mnt.0/': Permission denied
```

Reply:
The state of permission bits on the directory containing the mountpoint seems to be the troublemaker -

```
$ ls -la /tmp/tmpL8d3v_/
total 1
drwx------.  3 rishabh rishabh   60 Mar 25 17:04 .
drwxrwxrwt. 33 root    root    1740 Mar 25 17:04 ..
drwxrwxrwt.  2 rishabh rishabh    0 Mar 25 17:04 mnt.0
```

Fixing that fixes the issue with "file group" tests too.

Reply:
Glad you figured it out.